internal/crypto/srp.go

```diff
@@ -0,0 +1,248 @@
+package crypto
+
+import (
+	"crypto/sha512"
+	"io"
+	"math/big"
+
+	"golang.org/x/crypto/pbkdf2"
+	"golang.org/x/xerrors"
+
+	"github.com/gotd/xor"
+)
+
+type SRP struct {
+	random io.Reader
+}
+
+func NewSRP(random io.Reader) SRP {
+	return SRP{random: random}
+}
+
+type SRPInput struct {
+	// One of two salts used by the derivation function (see SRP 2FA login)
+	Salt1 []byte
+	// One of two salts used by the derivation function (see SRP 2FA login)
+	Salt2 []byte
+	// Base (see SRP 2FA login)
+	G int
+	// 2048-bit modulus (see SRP 2FA login)
+	P []byte
+}
+
+type SRPAnswer struct {
+	// A parameter (see SRP)
+	A []byte
+	// M1 parameter (see SRP)
+	M1 []byte
+}
+
+//nolint:gocritic
+func (s SRP) Auth(password, srpB, random []byte, i SRPInput) (SRPAnswer, error) {
```

Comment:
Probably we can move it to separate package, like `internal/crypto/srp` for some good naming?